pkg/clockcache/cache.go

```diff
@@ -0,0 +1,201 @@
+package clockcache
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+// CLOCK based approximate LRU storing mappings from strings to
+// int64s designed for concurrent usage.
+// It is sharded into 127 blocks, each of which can be
+// locked independently. Gets only require a read lock on the
+// individual block the element can be found in. When the cache is
+// not full, Insert only requires a write lock on the individual
+// block. Eviction locks blocks one at a time looking for a value
+// that's valid to evict/
+type Cache struct {
+	lock sync.RWMutex
```

Comment:
This lock protects access to `elements`+`storage` and `next`. I suggest having a separate Mutex for `storage`+`next`. This allows cache reads to take a read lock without having to wait for a pending eviction from a concurrent insert.

RWMutexes have a pathological case where every call to Lock causes all RLocks to be queued until Unlock is called.

Reply:
@iand you mean something like
```
1. lock storage+next, search for an element to be evicted
2. lock elements, evict old element and re-insert a new one
```
?
I'll have to think about that a bit more...

Reply:
one issue is that it can cause the evictor to starve if the other threads are hitting the cache fast enough...

Reply:
Yes, that occured to me while I was out for my walk. If the cache is hot such that every key is being used between writes then it's probably too small and inserting will just cause unnecessary churn.  However, is there a requirement for insert to always succeed? It could attempt one run through the eviction loop and return if it can't insert. On a hot cache that's not much different from inserting and being immediately evicted.

Reply:
That's a good point, and in the current implementation `InsertBatch` can actually "succeed" by evicting its own elements, which is morally the same thing.

Reply:
@iand @cevian Ok, let;s go through this in more depth.
The evict we use in `Insert` will go through `storage` until we either find an element to evict or we reach the element we started at without having done so. In the second case we’ll give up, as all the elements are too useful (maybe we’ll add a new heuristic later, really this probably means the cache is too small). I’m going to assume it’s ok to be a little sloppy with `used`, and that a sequence like
```
1. evictor sees Element e is unused, decides to evict it
2. getter marks e as used
3. evictor grabs lock on elements and evicts e
```
is ok: it’s morally equivalent to swapping the order of steps `1` and `2`.

The changed functions are
```go
func (self *Cache) insert(key interface{}, value interface{}) (canonicalKey interface{}, canonicalValue interface{}, inserted bool) {
	elem, present := self.elements[key]
	if present {
		return elem.key, elem.Value, false
	}

	var insertLocation *Element
	if len(self.storage) >= cap(self.storage) {
		insertLocation = self.evict()
		if insertLocation == nil {
			return nil, nil, false
		}
		self.elementsLock.Lock()
		defer self.elementsLock.Unlock()
		key := insertLocation.key
		delete(self.elements, key)
		*insertLocation = Element{key: key, Value: value}
	} else {
		self.storage = append(self.storage, Element{key: key, Value: value})
		insertLocation = &self.storage[len(self.storage)-1]
	}

	self.elements[key] = insertLocation
	return key, value, true
}

func (self *Cache) evict() (insertPtr *Element) {
	for {
		if self.next >= len(self.storage) {
			self.next = 0
		}

		insertPtr = self.tryEvictNext()
		if insertPtr != nil {
			return
		}
		if self.next == startLock {
			break
		}
	}
	// we've gone through all the elements, and they're all used
	// we'll try going through them once more, and if we cannot find
	// an unused element, the cache is to small, and we'll give up
	for {
		if self.next >= len(self.storage) {
			self.next = 0
		}

		insertPtr = self.tryEvictNext()
		if insertPtr != nil {
			return
		}
		if self.next == startLock {
			break
		}
	}
	return
}

func (self *Cache) tryEvictNext() (insertPtr *Element) {
	elem := &self.storage[self.next]
	used := atomic.CompareAndSwapUint32(&elem.used, 1, 0)
	if !used {
		insertPtr = elem
	}
	self.next += 1
	return
}

```

Reply:
so:

1. The long-duration lock only excludes other evictions, which is really nice; the only place we lock out `Get`s is for a `delete;insert` into the `map`
2.Correctness looks pretty trivial: the only state touched by the evictor is `next`, which is guarded by the lock, and `used` which is only accessed via atomics.
3. We can _almost_ forgo locking during the eviction-search entirely, if we use atomic ops to update `next` there're no race conditions, though there is a little weirdness when to evictors decide to evict the same element if one loops the other. I'm still tempted tho, I think that case should be rare enough, and the extra performance would be worth it. (this can be mitigated to a large degree by css-ing `2` into `used` when deciding to evict, it's not perfect because the `Get` update of `used` is non-atomic, but I'm not sure we care)
3. CPU-cache behavior is roughly the same: everyone has an oblige cache-miss to grab the lock, the evictor has an obligate miss per-element searched through, and however much it does when manipulating the map; the only addition is an extra lock-acquisition

Reply:
@iand nice find! if you and @cevian can confirm that the above code looks right (modulo the lock changes in `Insert[Batch]` which I elided) I'll update the PR to match

Reply:
writing this down so I don't forget: non-locking eviction has an actual data-race where the
```go
*insertLocation = Element{key: key, Value: value}
```
races against 
```go
used := atomic.CompareAndSwapUint32(&elem.used, 1, 0)
```
I think this is "benign" in that both are writing the same value, and I don't know of an architecture that would mess it up, but we still shouldn't do it.

Reply:
A few questions:
1) what lock protects next? I don't see it in the code above.
2) why is the Get update to Used non-atomic?

Reply:
Also what are the locks taken by Get?

Reply:
@cevian 
> what lock protects next? I don't see it in the code above.
Yeah, that's the one I elided 😊 . There's a mutex that protects next, and is acquired [here](https://github.com/timescale/timescale-prometheus/pull/147/files#diff-666b355b2c2f901c37b3cccc26156de4R52-R53l) and [here](https://github.com/timescale/timescale-prometheus/pull/147/files#diff-666b355b2c2f901c37b3cccc26156de4R67-R68) instead of the current lock.

> why is the Get update to Used non-atomic?
we don't have enough words. It's non-atomic because it uses
```go
if atomic.LoadUint32(&elem.used) == 0 {
    atomic.StoreUint32(&elem.used, 0)
}
```
instead of 
```go
atomic.CompareAndSwap(...)
```
for reasons described [here](https://github.com/timescale/timescale-prometheus/pull/147/files#r453666623), and that means the `insert` update to `used` can occur between the `if` and the `Store`. (It still is atomic in the sense that it uses atomic operations to shared memory, and does not have data-races)

Reply:
> Also what are the locks taken by Get?

@cevian The exact same locks that are taken now; the `Get` code should not need to change.

Reply:
This looks better to me